### PR TITLE
REF: use fused types for join_helper

### PR DESCRIPTION
**Pull request by jbrockmendel**
Related but orthogonal to #23022.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
I'm seeing this same Travis failure in several unrelated PRs.

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/issues/23172

I haven't made any progress debugging it though.

File: pandas/core/indexes/numeric.py [hunk 1]
```diff
@@ -321,13 +341,22 @@ class Float64Index(NumericIndex):
...
-    _left_indexer_unique = libjoin.left_join_indexer_unique_float64
-    _left_indexer = libjoin.left_join_indexer_float64
-    _inner_indexer = libjoin.inner_join_indexer_float64
-    _outer_indexer = libjoin.outer_join_indexer_float64
-
     _default_dtype = np.float64
 
+    # Cython methods; see github.com/cython/cython/issues/2647
+    #  for why we need to wrap these instead of making them class attributes
+    def _left_indexer_unique(self, left, right):
```
Comment (jreback) [inline pandas/core/indexes/numeric.py]: we could think about doing generic dispatch in the base class,
e.g.
```
def _left_indexer(self, left, right):
    return getattr(libjoin, 'left_join_indexer_{}.format(self._join_type))(left, right)
```

might be slightly less readable, but less code overall in the subclasses.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/join.pyx [hunk 2]
```diff
@@ -239,4 +239,420 @@ def ffill_indexer(ndarray[int64_t] indexer):
...
+
+left_join_indexer_unique_float64 = left_join_indexer_unique["float64_t"]
+left_join_indexer_unique_float32 = left_join_indexer_unique["float32_t"]
+left_join_indexer_unique_object = left_join_indexer_unique["object"]
+left_join_indexer_unique_int32 = left_join_indexer_unique["int32_t"]
+left_join_indexer_unique_int64 = left_join_indexer_unique["int64_t"]
+left_join_indexer_unique_uint64 = left_join_indexer_unique["uint64_t"]
+
+
+# @cython.wraparound(False)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/_libs/join.pyx]: Maintaining the status quo.

File: pandas/core/indexes/numeric.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/numeric.py]: Actually we can do one better.  By just calling `left_join_indexer` cython will dispatch to the appropriate version on its own.  (according to the docs; I haven't tried this myself)

**Comment by jreback** [comment]
thanks!

**Pull request merged**